### ISSUE TITLE
Enhancement: Templating engine

### ISSUE THREAD

**Issue by DakotaNelson** [EmpireProject/Empire#850]
I'm about to find myself with a lot of free time next week, and want to take a look to see if this would be worth me taking on:

### Proposal
Currently, a lot of code generation (for stagers, launchers, etc.) looks like:

```
stager = 'some stuff'
stager += 'more stuff'
stager += 'even more stuff'
if foo:
    stager += 'conditional stuff'
```

Maybe it's just me, but this seems to lead to some problems:
- makes generation logic hard to read (and thus debug)
- causes a lot of repetition (for instance, the first 700-ish lines of [the dropbox stager](https://github.com/EmpireProject/Empire/blob/master/data/agent/stagers/dropbox.py) and [the http stager](https://github.com/EmpireProject/Empire/blob/master/data/agent/stagers/http.py) are identical except for some weird whitespace differences - anyone wanting to make a new stager needs to copy/paste >700 lines of code)
- doesn't separate Python code from "other code" being generated, which makes working on the "other code" difficult, both in terms of development loop (change -> run empire -> generate -> debug) and tooling (syntax highlighting, debugging tools, etc. don't work on code embedded as a string in Python)

I think a good solution would be to add a templating engine to Empire, so that code generation can be accomplished by templating the necessary values into a standalone template file rather than building the template as we go using `+=`.

This is already done in some parts of Empire. For instance, it could work like seen here:
https://github.com/EmpireProject/Empire/blob/c95cc7cddd1f44705a2e89ea4f108b2b68f293ac/lib/listeners/http.py#L510-L522

But would instead use a templating engine to look more like:

`data/agent/stagers/http.py`:
```python
{% extends "base-stager.py" %}
{# The previous line will allow a "base stager" that contains any shred code between all stagers. #}
{# Also - this is a comment and so won't be included in the generated code. #}

WorkingHours = {{ working_hours }}
{# etc. #}
```

`lib/listeners/http.py`:
```python
from jinja2 import Environment, FileSystemLoader
# tell jinja2 where to get templates from (enables template inheritance)
env = Environment(
    loader=FileSystemLoader('/data/agent/stagers')
)

if host.endswith("/"):
    host = host[0:-1]

template = env.get_template('http.py')
return template.render(host=host, working_hours=working_hours)
```

This means that the stagers has less (no) repeated code since it can inherit from a stager base, the file is easier to read (and doesn't have "REPLACE ME" values), and the code which handles templating can be much easier (no `strip_python_comments`, for example). In addition, advanced features such as using for loops within templates becomes possible. (or even just sections of generated code which are included or not based on a boolean passed during generation)

#### Where do we keep the templates?
In `/data/` as is done now - there are already a bunch of templates there.

#### Won't this be a ton of work, to convert everything over?
It can be done bit by bit - there's no need to convert everything now. Files which use the templating engine will just have template artefacts instead of "REPLACE ME" values. It might be worth having a comment at the top of these files to delineate them, but they'll basically be the same.

In essence, I'd like to build the helper functions and other primitives to make this possible (and convert a few things as examples), but it's up to the module developers if they want to use it at the end of the day.

#### Why [jinja2](http://jinja.pocoo.org/)?

Two reasons:
1. It has logic
2. I've worked with it and like it :)

That said, perhaps a logicless system like [mustache](http://mustache.github.io/) would be better - or just a different system that does have logic. ([there are a lot](https://wiki.python.org/moin/Templating)) I'm 100% open to feedback here.

Anyway, the point here is that I think this is worth doing and want to check to see if A) this is worth doing at all and B) it's worth doing like this. If the answer to both is approximately truthy I'll go ahead and get started next week.

Thoughts?

**Comment by xorrior** [comment]
@DakotaNelson Please have at it, this sounds like an amazing idea. As the project continues to grow, it's become even harder for maintenance and new additions. I would love to shrink the codebase down and this sounds like it's a solid answer.

**Comment by xorrior** [comment]
Merged #863. Thanks again!
